src/lib/sqle/connection.service.ts

```diff
@@ -1,74 +1,223 @@
+    connection: ISQLEConnection,
+    options?: IConnectOptions,
+  ): Observable<ISQLEConnection> {
+    if (this._getConnectionIndex(connection) > -1) {
+      return this._pingAndSave(connection, true, options);
+    } else {
+      throw Error('Connection does not exist');
+    }
+  }
+  public removeConnection(connection: ISQLEConnection): ISQLEConnection {
+    const index: number = this._getConnectionIndex(connection);
+    if (index > -1) {
+      this._connections = [...this._connections.slice(0, index), ...this._connections.slice(index + 1)];
+      this._currentConnection =
+        this._currentConnection && this._connectionsAreEqual(this._currentConnection, connection)
+          ? undefined
+          : this._currentConnection;
+      return connection;
+    } else {
+      // connection does not exist but that is fine?
+      return undefined;
+    }
+  }
+
+  public disconnectFromCurrentConnection(): void {
+    this._currentConnection = undefined;
+  }
+
+  public removeAllConnections(): void {
+    this._connections = [];
+    this._currentConnection = undefined;
+  }
+
+  public connectionExits(connection: ISQLEConnection): boolean {
+    return this._getConnectionIndex(connection) > -1;
+  }
+
+  public connectionIsCurrentConnection(connection: ISQLEConnection): boolean {
+    return this._connectionsAreEqual(connection, this.currentConnection);
   }
 
-  public disconnect(): void {
-    sessionStorage.removeItem(CONNECTION_SESSION_KEY);
+  public stringifyConnection(connection: ISQLEConnection): string {
+    return stringifyConnection(connection);
   }
 
-  public connect(
+  private _pingAndSave(
     connection: ISQLEConnection,
-    opts?: { timeout: number; attempts: number },
+    setAsCurrent: boolean,
+    opts: IConnectOptions = { attempts: 2, timeout: 7000 },
   ): Observable<ISQLEConnection> {
-    // clear connection before starting a new one
-    this.disconnect();
     // test connection with SELECT 1
     return this._queryService.querySystem(connection, { query: 'SELECT 1;' }).pipe(
       // timeout connection if more than 7 seconds
-      timeout(opts?.timeout || 7000),
+      timeout(opts.timeout),
       // retry only after a certain number of attempts or if the error is something else than 420
       retryWhen((errors: Observable<{ httpStatus: number }>) => {
         return errors.pipe(
           mergeMap((error: { httpStatus: number }, index: number) => {
             const retryAttempt: number = index + 1;
-            if (retryAttempt > (opts?.attempts || 2) || error.httpStatus === 420) {
+            if (retryAttempt > opts.attempts || error.httpStatus === 420) {
               return throwError(error);
             }
             return timer(0);
           }),
         );
       }),
-      tap(() => this.store(connection)), // if successful, save
+      tap(() => {
+        // if successful, save
+        const index: number = this._getConnectionIndex(connection);
```

Comment:
Is this just to verify the connection doesnt exist already?

Should we just verify that at the beginning and throw an exception?

Reply:
Errors should already be thrown by the methods that call `_pingAndSave`